 **Rental Bike Sharing**


---
---


## Carga de datos

fuente:https://www.kaggle.com/datasets/imakash3011/rental-bike-sharing



In [ ]:
from google.colab import drive
import os
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as mpl
import matplotlib.pyplot as plt


drive.mount('/content/drive')
%cd  "/content/drive/MyDrive/Coderhouse/Desafios"


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
/content/drive/MyDrive/Coderhouse/Desafios


In [ ]:
df = pd.read_csv ("Bicicletas por hora.csv",sep = ",")

In [ ]:
df.head() 


instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   
1        2  2011-01-01       1   0     1   1        0        6           0   
2        3  2011-01-01       1   0     1   2        0        6           0   
3        4  2011-01-01       1   0     1   3        0        6           0   
4        5  2011-01-01       1   0     1   4        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16  
1           1  0.22  0.2727  0.80        0.0       8          32   40  
2           1  0.22  0.2727  0.80        0.0       5          27   32  
3           1  0.24  0.2879  0.75        0.0       3          10   13  
4           1  0.24  0.2879  0.75        0.0       0           1    1

De acuerdo a las entregas anteriores, cnt (cantidad de bicicletas alquiladas) varia segun la hora (hr), categoria de clima (weathersit)

# Regresión lineal


## Segun categoria de clima y horas

In [ ]:
X = df[["weathersit","hr"]]
y = df.cnt
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split (X,y,test_size = .2,random_state = 1)
X_train

weathersit  hr
9088            1  14
1292            1   0
13559           1   4
4145            1  15
15913           2  18
...           ...  ..
10955           1  15
17289           1   6
5192            1   6
12172           1   9
235             2   2

[13903 rows x 2 columns]

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error , mean_squared_error

lr = LinearRegression ()
lr.fit (X_train, y_train)
y_pred = lr.predict(X_test)

print("R2", r2_score(y_test,y_pred))
print("mae", mean_absolute_error(y_test,y_pred))
print("mse", mean_squared_error(y_test,y_pred))
print("rmse", (mean_squared_error(y_test,y_pred))**(1/2))


R2 0.18252899711255766
mae 123.76044274410856
mse 27272.400134816795
rmse 165.14357430677342


## Según componentes del clima y horas

In [ ]:
X = df[["temp","atemp","hum","windspeed","hr"]]
y = df.cnt
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split (X,y,test_size = .2,random_state = 1)


In [ ]:
lr = LinearRegression ()
lr.fit (X_train, y_train)
y_pred = lr.predict(X_test)

print("R2", r2_score(y_test,y_pred))
print("mae", mean_absolute_error(y_test,y_pred))
print("mse", mean_squared_error(y_test,y_pred))
print("rmse", (mean_squared_error(y_test,y_pred))**(1/2))

R2 0.33299128911724674
mae 109.19924209368044
mse 22252.6895661735
rmse 149.17335407563075


## Segun compomentes del clima, hora, dia de la semana y dia laboral

In [ ]:
X = df[["temp","atemp","hum","windspeed","hr","weekday","workingday"]]
y = df.cnt
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split (X,y,test_size = .2,random_state = 1)


In [ ]:
lr = LinearRegression ()
lr.fit (X_train, y_train)
y_pred = lr.predict(X_test)

print("R2", r2_score(y_test,y_pred))
print("mae", mean_absolute_error(y_test,y_pred))
print("mse", mean_squared_error(y_test,y_pred))
print("rmse", (mean_squared_error(y_test,y_pred))**(1/2))

R2 0.33341955061873085
mae 109.3325667959492
mse 22238.40194130417
rmse 149.1254570531275


por ahora, esta es la mejor opcion, pero el r2 sigue siendo muy bajo

In [ ]:
df.describe()


instant        season            yr          mnth            hr  \
count  17379.0000  17379.000000  17379.000000  17379.000000  17379.000000   
mean    8690.0000      2.501640      0.502561      6.537775     11.546752   
std     5017.0295      1.106918      0.500008      3.438776      6.914405   
min        1.0000      1.000000      0.000000      1.000000      0.000000   
25%     4345.5000      2.000000      0.000000      4.000000      6.000000   
50%     8690.0000      3.000000      1.000000      7.000000     12.000000   
75%    13034.5000      3.000000      1.000000     10.000000     18.000000   
max    17379.0000      4.000000      1.000000     12.000000     23.000000   

            holiday       weekday    workingday    weathersit          temp  \
count  17379.000000  17379.000000  17379.000000  17379.000000  17379.000000   
mean       0.028770      3.003683      0.682721      1.425283      0.496987   
std        0.167165      2.005771      0.465431      0.639357      0.192556   
min        0.000000      0.000000      0.000000      1.000000      0.020000   
25%        0.000000      1.000000      0.000000      1.000000      0.340000   
50%        0.000000      3.000000      1.000000      1.000000      0.500000   
75%        0.000000      5.000000      1.000000      2.000000      0.660000   
max        1.000000      6.000000      1.000000      4.000000      1.000000   

              atemp           hum     windspeed        casual    registered  \
count  17379.000000  17379.000000  17379.000000  17379.000000  17379.000000   
mean       0.475775      0.627229      0.190098     35.676218    153.786869   
std        0.171850      0.192930      0.122340     49.305030    151.357286   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.333300      0.480000      0.104500      4.000000     34.000000   
50%        0.484800      0.630000      0.194000     17.000000    115.000000   
75%        0.621200      0.780000      0.253700     48.000000    220.000000   
max        1.000000      1.000000      0.850700    367.000000    886.000000   

                cnt  
count  17379.000000  
mean     189.463088  
std      181.387599  
min        1.000000  
25%       40.000000  
50%      142.000000  
75%      281.000000  
max      977.000000

## Todas las variables 

In [ ]:
X = df[["temp","atemp","hum","windspeed","hr","weekday","workingday","holiday","season","weathersit"]]
y = df.cnt
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split (X,y,test_size = .2,random_state = 1)

In [ ]:
lr = LinearRegression ()
lr.fit (X_train, y_train)
y_pred = lr.predict(X_test)

print("R2", r2_score(y_test,y_pred))
print("mae", mean_absolute_error(y_test,y_pred))
print("mse", mean_squared_error(y_test,y_pred))
print("rmse", (mean_squared_error(y_test,y_pred))**(1/2))

R2 0.3449788305559731
mae 108.7566381499657
mse 21852.762197991782
rmse 147.82679796975845


## Conclusión:

**Como el objetivo es precedir la cantidad de alquileres, lo optimo es hacer una regresión**

**En todos los casos, el R2 es muy bajo, por lo que entiendo que la regresión lineal no es adecuada para este dataset**

**al agregar todas las variables se obtuvo el menor error, pero el rmse da un desvio promedio de 147 alquileres, cuando el promedio de los alquileres es de 189.**

**No parece ser un modelo eficiente de predicción**